File: rad/engine/02_catalog/store/transition_compaction.go
package store

import (
	"context"
	"fmt"
	"time"

	kv "github.com/Southclaws/rad/rad/engine/01_kv"
	"github.com/Southclaws/rad/rad/engine/02_catalog/model"
)

// CompactTransition discards worker-only state after physical cleanup while
// preserving the stable control result and terminal diagnostics at the same
// transition identity.
func CompactTransition(ctx context.Context, view kv.KV, id string, completedAt time.Time) error {
	transition, ok, err := GetTransition(ctx, view, id)
	if err != nil {
		return err
	}
	if !ok {
		return fmt.Errorf("catalog: transition %q does not exist", id)
	}
	switch transition.State {
	case model.TransitionReady, model.TransitionCancelled, model.TransitionFailed:
	default:
		return fmt.Errorf("catalog: transition %q is not terminal", id)
	}
	if !transition.CompactedAt.IsZero() {
		return nil
	}
	transition.Index = model.Index{
		ID:                   transition.Index.ID,
		LogicalID:            transition.Index.LogicalID,
		DefinitionGeneration: transition.Index.DefinitionGeneration,
		State:                transition.Index.State,
		Name:                 transition.Index.Name,
		Unique:               transition.Index.Unique,
	}
	transition.OwnerEpoch = 0
	transition.BasePosition = ""
	transition.BarrierPosition = ""
	transition.TableID = ""
	transition.Cursor = nil
	transition.BatchID = 0
	transition.AppliedDelta = 0
	transition.DeltaHighWater = 0
	transition.DeltaSoftLimit = 0
	transition.DeltaHardLimit = 0
	transition.WorkState = ""
	transition.RowsScanned = 0
	transition.AffectedColumnIDs = nil
	transition.IndexRequest = nil
	transition.ColumnReplacement = nil
	transition.ReplacementRequest = nil
	transition.Constraint = nil
	transition.ConstraintRequest = nil
	transition.Prerequisites = nil
	transition.GateTableIDs = nil
	transition.CompactedAt = completedAt.UTC()
	return SaveTransition(ctx, view, transition)
}

File: src/engine/02_catalog/store/transition_compaction.rs
use crate::engine::catalog::identity::{AccessGeneration, OwnerEpoch, TableId};
use crate::engine::catalog::model::{
    DataPosition, Index, Timestamp, TransitionState, TransitionWorkState,
};
use crate::engine::catalog::{Error, ErrorKind, Result};
use crate::engine::kv::KvView;

use super::{get_transition, save_transition};
use crate::engine::catalog::identity::TransitionId;

pub async fn compact_transition<V: KvView + ?Sized>(
    view: &mut V,
    id: &TransitionId,
    completed_at: Timestamp,
) -> Result<()> {
    let Some(mut transition) = get_transition(view, id).await? else {
        return Err(Error::message(
            ErrorKind::NotFound,
            format!("catalog: transition {id:?} does not exist"),
        ));
    };
    if !matches!(
        transition.state,
        TransitionState::Ready | TransitionState::Cancelled | TransitionState::Failed
    ) {
        return Err(Error::message(
            ErrorKind::InvalidInput,
            format!("catalog: transition {id:?} is not terminal"),
        ));
    }
    if !transition.compacted_at.is_zero() {
        return Ok(());
    }
    transition.index = Index {
        id: transition.index.id,
        logical_id: transition.index.logical_id,
        definition_generation: transition.index.definition_generation,
        access_generation: AccessGeneration::ZERO,
        state: transition.index.state,
        name: transition.index.name,
        columns: Vec::new(),
        column_ids: Vec::new(),
        unique: transition.index.unique,
    };
    transition.owner_epoch = OwnerEpoch::ZERO;
    transition.base_position = DataPosition::default();
    transition.barrier_position = DataPosition::default();
    transition.table_id = TableId::default();
    transition.cursor.clear();
    transition.batch_id = 0;
    transition.applied_delta = 0;
    transition.delta_high_water = 0;
    transition.delta_soft_limit = 0;
    transition.delta_hard_limit = 0;
    transition.work_state = TransitionWorkState::Normal;
    transition.rows_scanned = 0;
    transition.affected_column_ids.clear();
    transition.index_request = None;
    transition.column_replacement = None;
    transition.replacement_request = None;
    transition.constraint = None;
    transition.constraint_request = None;
    transition.prerequisites.clear();
    transition.gate_table_ids.clear();
    transition.compacted_at = completed_at;
    save_transition(view, &transition).await
}
